### ISSUE TITLE
Actor's animation freezes while moving if he's facing the same direction as his movement

### ISSUE THREAD

**Issue by nagman** [chrismaltby/gb-studio#915]
**Describe the bug**
If an actor is facing the direction he'll be moving, he's not animated while moving.

Here's a video that demonstrates it:

https://user-images.githubusercontent.com/16116486/146841256-81140e48-6c5b-4e31-9e83-2f9c0c6b7464.mp4

The framerate is a little janky, but you can see that the actor only moves his legs when he's not facing north in the first place.

**Platform (please complete the following information):**
 - macOS 11.6
 - App version: 3.0.1

**Comment by Rebusmind** [comment]
Duplicate of this issue (but still not fixed apparently): https://github.com/chrismaltby/gb-studio/issues/787

**Comment by nagman** [comment]
Still not fixed with GB Studio 3.0.2

**Comment by Rebusmind** [comment]
I might take a look myself after Christmas.

**Comment by reakain** [comment]
Alright, I think I found the issue. In the `vm_actor_move_to` function in `vm_actor.c` never sets the animation set to the walking animations data, so it's still using the idle set, which is just the four directions. Changing the direction works because in that function it does a check to see if the direction is new, and if it is, then it sets the spritesheet to the new direction, but in that call has the "moving" flag as True, so it gets the animation data then. I added a call to set the spritesheet to the moving one where it sets the actor flag to moving, and that fixed the issue. 

I don't know if this introduces any new and exciting issues. This wont fix the trigger issue, as that is the inverse. I think a more complete fix would be something where setting/unsetting the movement flag automatically flips the animation state, and having those trigger "pause" actions flip the movement flags until complete... I don't feel particularly well equipped to handle making a larger fix like that though....

It looks kind of like with the current implementation, you could have the idle and moving framesets be completely separate, and only import one or the other, but that's not what it's doing either. It's either importing idle only, or idle and movement. Was this maybe the original method that was implemented to fix the "walking while standing still" behavior from the early version?

Tl;dr, I added `actor_set_anim_moving(actor);` line after line 54 in `vm_actor.c`  which amounts to added an animation sheet check after setting the move_active flag.

Function with line:
```
void vm_actor_move_to(SCRIPT_CTX * THIS, INT16 idx) OLDCALL BANKED {
    actor_t *actor;
    direction_e new_dir = DIR_DOWN;

    // indicate waitable state of context
    THIS->waitable = 1;

    act_move_to_t * params = VM_REF_TO_PTR(idx);
    actor = actors + params->ID;

    if (THIS->flags == 0) {
        THIS->flags = MOVE_ACTIVE;
        actor_set_anim_moving(actor);    // <------------- This is the bit I added

        // Snap to nearest pixel before moving
        actor->pos.x = ((actor->pos.x >> 4) << 4);
        actor->pos.y = ((actor->pos.y >> 4) << 4);

        if (CHK_FLAG(params->ATTR, ACTOR_ATTR_DIAGONAL)) {
            SET_FLAG(THIS->flags, MOVE_ALLOW_H | MOVE_ALLOW_V);
        } if (CHK_FLAG(params->ATTR, ACTOR_ATTR_H_FIRST)) {
            if (actor->pos.x != params->X) {
                SET_FLAG(THIS->flags, MOVE_ALLOW_H);
            } else {
                SET_FLAG(THIS->flags, MOVE_ALLOW_V);
            }            
        } else {
            if (actor->pos.y != params->Y) {
                SET_FLAG(THIS->flags, MOVE_ALLOW_V);
            } else {
                SET_FLAG(THIS->flags, MOVE_ALLOW_H);
            }
        }

        // Check for collisions in path
        if (CHK_FLAG(params->ATTR, ACTOR_ATTR_CHECK_COLL)) {
            if (CHK_FLAG(params->ATTR, ACTOR_ATTR_H_FIRST)) {
                // Check for horizontal collision
                if (actor->pos.x != params->X) {
                    UBYTE check_dir = (actor->pos.x > params->X) ? CHECK_DIR_LEFT : CHECK_DIR_RIGHT;
                    params->X = check_collision_in_direction(actor->pos.x, actor->pos.y, &actor->bounds, params->X, check_dir);
                }
                // Check for vertical collision
                if (actor->pos.y != params->Y) {
                    UBYTE check_dir = (actor->pos.y > params->Y) ? CHECK_DIR_UP : CHECK_DIR_DOWN;
                    params->Y = check_collision_in_direction(params->X, actor->pos.y, &actor->bounds, params->Y, check_dir);
                }
            } else {
                // Check for vertical collision
                if (actor->pos.y != params->Y) {
                    UBYTE check_dir = (actor->pos.y > params->Y) ? CHECK_DIR_UP : CHECK_DIR_DOWN;
                    params->Y = check_collision_in_direction(actor->pos.x, actor->pos.y, &actor->bounds, params->Y, check_dir);
                }
                // Check for horizontal collision
                if (actor->pos.x != params->X) {
                    UBYTE check_dir = (actor->pos.x > params->X) ? CHECK_DIR_LEFT : CHECK_DIR_RIGHT;
                    params->X = check_collision_in_direction(actor->pos.x, params->Y, &actor->bounds, params->X, check_dir);
                }
            }
        }
    }

    // Actor reached destination
    if ((actor->pos.x == params->X) && (actor->pos.y == params->Y)) {
        THIS->flags = MOVE_INACTIVE;
        actor_set_anim_idle(actor);
        return;
    }

    // Move in X Axis
    if (CHK_FLAG(THIS->flags, MOVE_ALLOW_H) && (actor->pos.x != params->X)) {
        if (actor->pos.x < params->X) {
            new_dir = DIR_RIGHT;
        } else if (actor->pos.x > params->X) {
            new_dir = DIR_LEFT;
        }
        
        // Move actor
        point_translate_dir(&actor->pos, new_dir, actor->move_speed);

        // Check for actor collision
        if (CHK_FLAG(params->ATTR, ACTOR_ATTR_CHECK_COLL) && !CHK_FLAG(THIS->flags, MOVE_ALLOW_V) && actor_overlapping_bb(&actor->bounds, &actor->pos, actor, FALSE)) {
            point_translate_dir(&actor->pos, FLIPPED_DIR(new_dir), actor->move_speed);   
            THIS->flags = 0;
            actor_set_anim_idle(actor);
            return;
        }

        // Check if overshot destination
        if (new_dir == DIR_LEFT &&  (actor->pos.x < params->X)) {
            actor->pos.x = params->X;
        } else if (new_dir == DIR_RIGHT && (actor->pos.x > params->X)) {
            actor->pos.x = params->X;
        }

        // Reached Horizontal Destination
        if (actor->pos.x == params->X) {
            SET_FLAG(THIS->flags, MOVE_ALLOW_V);
            CLR_FLAG(THIS->flags, MOVE_ALLOW_H);
        }        
    }

    // Move in Y Axis
    if (CHK_FLAG(THIS->flags, MOVE_ALLOW_V) && (actor->pos.y != params->Y)) {
        if (actor->pos.y < params->Y) {
            new_dir = DIR_DOWN;
        } else if (actor->pos.y > params->Y) {
            new_dir = DIR_UP;
        }
        
        // Move actor
        point_translate_dir(&actor->pos, new_dir, actor->move_speed);

        // Check for actor collision
        if (CHK_FLAG(params->ATTR, ACTOR_ATTR_CHECK_COLL) && actor_overlapping_bb(&actor->bounds, &actor->pos, actor, FALSE)) {
            point_translate_dir(&actor->pos, FLIPPED_DIR(new_dir), actor->move_speed);   
            THIS->flags = 0;
            actor_set_anim_idle(actor);
            return;
        }

        // Check if overshot destination
        if (new_dir == DIR_UP && (actor->pos.y < params->Y)) {
            actor->pos.y = params->Y;
        } else if (new_dir == DIR_DOWN &&  (actor->pos.y > params->Y)) {
            actor->pos.y = params->Y;
        }

        // Reached Vertical Destination
        if (actor->pos.y == params->Y) {
            SET_FLAG(THIS->flags, MOVE_ALLOW_H);
            CLR_FLAG(THIS->flags, MOVE_ALLOW_V);
        }
    }

    // If changed direction, trigger actor rerender
    if (actor->dir != new_dir) {
        actor_set_dir(actor, new_dir, TRUE);        
    }

    THIS->PC -= (INSTRUCTION_SIZE + sizeof(idx));
    return;
}
```

**Comment by maxoakland** [comment]
> Alright, I think I found the issue. In the `vm_actor_move_to` function in `vm_actor.c` never sets the animation set to the walking animations data, so it's still using the idle set, which is just the four directions. Changing the direction works because in that function it does a check to see if the direction is new, and if it is, then it sets the spritesheet to the new direction, but in that call has the "moving" flag as True, so it gets the animation data then. I added a call to set the spritesheet to the moving one where it sets the actor flag to moving, and that fixed the issue.
> 
> I don't know if this introduces any new and exciting issues. This wont fix the trigger issue, as that is the inverse. I think a more complete fix would be something where setting/unsetting the movement flag automatically flips the animation state, and having those trigger "pause" actions flip the movement flags until complete... I don't feel particularly well equipped to handle making a larger fix like that though....
> 
> It looks kind of like with the current implementation, you could have the idle and moving framesets be completely separate, and only import one or the other, but that's not what it's doing either. It's either importing idle only, or idle and movement. Was this maybe the original method that was implemented to fix the "walking while standing still" behavior from the early version?
> 
> Tl;dr, I added `actor_set_anim_moving(actor);` line after line 54 in `vm_actor.c` which amounts to added an animation sheet check after setting the move_active flag.
> 
> Function with line:
> 
> ```
> void vm_actor_move_to(SCRIPT_CTX * THIS, INT16 idx) OLDCALL BANKED {
>     actor_t *actor;
>     direction_e new_dir = DIR_DOWN;
> 
>     // indicate waitable state of context
>     THIS->waitable = 1;
> 
>     act_move_to_t * params = VM_REF_TO_PTR(idx);
>     actor = actors + params->ID;
> 
>     if (THIS->flags == 0) {
>         THIS->flags = MOVE_ACTIVE;
>         actor_set_anim_moving(actor);    // <------------- This is the bit I added
> 
>         // Snap to nearest pixel before moving
>         actor->pos.x = ((actor->pos.x >> 4) << 4);
>         actor->pos.y = ((actor->pos.y >> 4) << 4);
> 
>         if (CHK_FLAG(params->ATTR, ACTOR_ATTR_DIAGONAL)) {
>             SET_FLAG(THIS->flags, MOVE_ALLOW_H | MOVE_ALLOW_V);
>         } if (CHK_FLAG(params->ATTR, ACTOR_ATTR_H_FIRST)) {
>             if (actor->pos.x != params->X) {
>                 SET_FLAG(THIS->flags, MOVE_ALLOW_H);
>             } else {
>                 SET_FLAG(THIS->flags, MOVE_ALLOW_V);
>             }            
>         } else {
>             if (actor->pos.y != params->Y) {
>                 SET_FLAG(THIS->flags, MOVE_ALLOW_V);
>             } else {
>                 SET_FLAG(THIS->flags, MOVE_ALLOW_H);
>             }
>         }
> 
>         // Check for collisions in path
>         if (CHK_FLAG(params->ATTR, ACTOR_ATTR_CHECK_COLL)) {
>             if (CHK_FLAG(params->ATTR, ACTOR_ATTR_H_FIRST)) {
>                 // Check for horizontal collision
>                 if (actor->pos.x != params->X) {
>                     UBYTE check_dir = (actor->pos.x > params->X) ? CHECK_DIR_LEFT : CHECK_DIR_RIGHT;
>                     params->X = check_collision_in_direction(actor->pos.x, actor->pos.y, &actor->bounds, params->X, check_dir);
>                 }
>                 // Check for vertical collision
>                 if (actor->pos.y != params->Y) {
>                     UBYTE check_dir = (actor->pos.y > params->Y) ? CHECK_DIR_UP : CHECK_DIR_DOWN;
>                     params->Y = check_collision_in_direction(params->X, actor->pos.y, &actor->bounds, params->Y, check_dir);
>                 }
>             } else {
>                 // Check for vertical collision
>                 if (actor->pos.y != params->Y) {
>                     UBYTE check_dir = (actor->pos.y > params->Y) ? CHECK_DIR_UP : CHECK_DIR_DOWN;
>                     params->Y = check_collision_in_direction(actor->pos.x, actor->pos.y, &actor->bounds, params->Y, check_dir);
>                 }
>                 // Check for horizontal collision
>                 if (actor->pos.x != params->X) {
>                     UBYTE check_dir = (actor->pos.x > params->X) ? CHECK_DIR_LEFT : CHECK_DIR_RIGHT;
>                     params->X = check_collision_in_direction(actor->pos.x, params->Y, &actor->bounds, params->X, check_dir);
>                 }
>             }
>         }
>     }
> 
>     // Actor reached destination
>     if ((actor->pos.x == params->X) && (actor->pos.y == params->Y)) {
>         THIS->flags = MOVE_INACTIVE;
>         actor_set_anim_idle(actor);
>         return;
>     }
> 
>     // Move in X Axis
>     if (CHK_FLAG(THIS->flags, MOVE_ALLOW_H) && (actor->pos.x != params->X)) {
>         if (actor->pos.x < params->X) {
>             new_dir = DIR_RIGHT;
>         } else if (actor->pos.x > params->X) {
>             new_dir = DIR_LEFT;
>         }
>         
>         // Move actor
>         point_translate_dir(&actor->pos, new_dir, actor->move_speed);
> 
>         // Check for actor collision
>         if (CHK_FLAG(params->ATTR, ACTOR_ATTR_CHECK_COLL) && !CHK_FLAG(THIS->flags, MOVE_ALLOW_V) && actor_overlapping_bb(&actor->bounds, &actor->pos, actor, FALSE)) {
>             point_translate_dir(&actor->pos, FLIPPED_DIR(new_dir), actor->move_speed);   
>             THIS->flags = 0;
>             actor_set_anim_idle(actor);
>             return;
>         }
> 
>         // Check if overshot destination
>         if (new_dir == DIR_LEFT &&  (actor->pos.x < params->X)) {
>             actor->pos.x = params->X;
>         } else if (new_dir == DIR_RIGHT && (actor->pos.x > params->X)) {
>             actor->pos.x = params->X;
>         }
> 
>         // Reached Horizontal Destination
>         if (actor->pos.x == params->X) {
>             SET_FLAG(THIS->flags, MOVE_ALLOW_V);
>             CLR_FLAG(THIS->flags, MOVE_ALLOW_H);
>         }        
>     }
> 
>     // Move in Y Axis
>     if (CHK_FLAG(THIS->flags, MOVE_ALLOW_V) && (actor->pos.y != params->Y)) {
>         if (actor->pos.y < params->Y) {
>             new_dir = DIR_DOWN;
>         } else if (actor->pos.y > params->Y) {
>             new_dir = DIR_UP;
>         }
>         
>         // Move actor
>         point_translate_dir(&actor->pos, new_dir, actor->move_speed);
> 
>         // Check for actor collision
>         if (CHK_FLAG(params->ATTR, ACTOR_ATTR_CHECK_COLL) && actor_overlapping_bb(&actor->bounds, &actor->pos, actor, FALSE)) {
>             point_translate_dir(&actor->pos, FLIPPED_DIR(new_dir), actor->move_speed);   
>             THIS->flags = 0;
>             actor_set_anim_idle(actor);
>             return;
>         }
> 
>         // Check if overshot destination
>         if (new_dir == DIR_UP && (actor->pos.y < params->Y)) {
>             actor->pos.y = params->Y;
>         } else if (new_dir == DIR_DOWN &&  (actor->pos.y > params->Y)) {
>             actor->pos.y = params->Y;
>         }
> 
>         // Reached Vertical Destination
>         if (actor->pos.y == params->Y) {
>             SET_FLAG(THIS->flags, MOVE_ALLOW_H);
>             CLR_FLAG(THIS->flags, MOVE_ALLOW_V);
>         }
>     }
> 
>     // If changed direction, trigger actor rerender
>     if (actor->dir != new_dir) {
>         actor_set_dir(actor, new_dir, TRUE);        
>     }
> 
>     THIS->PC -= (INSTRUCTION_SIZE + sizeof(idx));
>     return;
> }
> ```

Great work! You could turn this into a pull request
